testdata/p4_16_samples_outputs/p416-type-use3.p4.p4info.txt

```diff
@@ -0,0 +1,216 @@
+pkg_info {
+  arch: "v1model"
+}
+tables {
+  preamble {
+    id: 33586815
+    name: "ingress.custom_table"
+    alias: "custom_table"
+  }
+  match_fields {
+    id: 1
+    name: "hdr.ethernet.srcAddr"
+    bitwidth: 48
+    match_type: EXACT
+    type_name {
+      name: "EthT_t"
+    }
+  }
+  match_fields {
+    id: 2
+    name: "hdr.ethernet.dstAddr"
+    bitwidth: 48
+    match_type: EXACT
+  }
+  match_fields {
+    id: 3
+    name: "hdr.custom.e"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 4
+    name: "hdr.custom.ed"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 5
+    name: "hdr.custom.et"
+    bitwidth: 8
+    match_type: EXACT
```

Comment:
I think this match_fields for "hdr.custom.et" should have this type:
```
type_name {
  name: "CustomT_t"
}
```
And similarly for many other match_fields (the ones that have a "t" in their name should, at least, but not the ones that have no "t"s in their name).

Reply:
I don't mean that they should all have the type named "CustomT_t", but all of the ones with names with "t" in them should have a type_name in the p4Info file.

Reply:
I should mention that I am OK with this being committed without this being fixed, as long as we create an issue to fix them in a future PR.

Reply:
None of the MatchFields have the p4runtime_translation annotation on them.  This is why the `type` does not represent them.  Please grep for "p4runtime_translation" in `p416-type-use3.p4` and see that only the ones that have an Annotation have been translated.

I just added the annotation above CustomT_t and then the type_names are added.

Reply:
The proposed rule for having a type_name field filled in is _not_ whether it is a custom `type` in the P4 source code that has a p4runtime_translation annotation -- it is whether it has a custom `type` in the P4 source code, whether it has that annotation or not.  Take a look at this write-up, which Antonin has now reviewed and believes is correct: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#proposed-p4c-rules-for-generating-types-in-p4info-files

The presence or absence of a p4runtime_translation annotation should affect where the bitwidth value comes from, and it should affect how the type is represented in the type_info message later, but type_name should be there for all table search key fields, value_set key fields, and action parameters, whenever they have a custom `type`.  (Later also for all controller packet in/out metadata fields, after this PR is merged in: https://github.com/p4lang/p4runtime/pull/188

Reply:
Thanks for clarifying.  I just checked in new code.  Please take a look at new p4info.txt files.

Reply:
I inspected the changed p4info.txt file and see type_name is being generated for any 'type` with a "T".